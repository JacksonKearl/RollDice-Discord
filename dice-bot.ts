import { ParserBuilder, ParenthesesParselet } from "./parse"
import { Tokenizer } from "./tokenize"
import {
  Name,
  Number,
  Roll,
  Add,
  Sub,
  Mul,
  Advantage,
  DiceExpression,
  Neg,
  Div,
  Disadvantage,
  Bang,
  Assign,
  ExpressionResult
} from "./dice-expression"
import { UserEnvironmentView } from "./environment"

export const { tokenize } = new Tokenizer([
  { pattern: "+=" },
  { pattern: "-=" },
  { pattern: "+" },
  { pattern: "-" },
  { pattern: "*" },
  { pattern: "/" },
  { pattern: "(" },
  { pattern: ")" },
  { pattern: "!" },
  { pattern: "=" },
  { pattern: /@\s*a(d(v(a(n(t(a(g(e)?)?)?)?)?)?)?)?\b/, name: "ADVANTAGE" },
  { pattern: /@\s*d(i(s(a(d(v(a(n(t(a(g(e)?)?)?)?)?)?)?)?)?)?)?\b/, name: "DISADVANTAGE" },
  { pattern: /(\d+)?(?:d|D)(\d+)(k\d+)?/, name: "ROLL" },
  { pattern: /[a-zA-Z](\w|\.)*/, name: "NAME" },
  { pattern: /\d+/, name: "NUMBER" }
])

const enum Precedence {
  Assign = 1,
  At = 2,
  AddSub = 3,
  MulDiv = 4,
  Bang = 5,
  Negate = 6
}

export const calculator = (env: UserEnvironmentView) =>
  new ParserBuilder<DiceExpression>()
    .registerPrefix("NUMBER", { parse: (_, token) => new Number(token.match) })
    .registerPrefix("ROLL", { parse: (_, token) => new Roll(token.match) })
    .registerPrefix("NAME", { parse: (_, token) => new Name(token.match, execute, env) })
    .registerPrefix("(", ParenthesesParselet)

    .prefix("-", Precedence.Negate, (_, right) => Neg(right))

    .postfix("!", Precedence.Bang, left => Bang(left))

    .postfix("ADVANTAGE", Precedence.At, left => Advantage(left))
    .postfix("DISADVANTAGE", Precedence.At, left => Disadvantage(left))

    .infixLeft("/", Precedence.MulDiv, (left, _, right) => Div(left, right))
    .infixLeft("*", Precedence.MulDiv, (left, _, right) => Mul(left, right))
    .infixLeft("+", Precedence.AddSub, (left, _, right) => Add(left, right))
    .infixLeft("-", Precedence.AddSub, (left, _, right) => Sub(left, right))

    .infixLeft("+=", Precedence.Assign, (left, _, right) =>
      Assign(left, Add(Bang(left), right), env)
    )
    .infixLeft("-=", Precedence.Assign, (left, _, right) =>
      Assign(left, Sub(Bang(left), right), env)
    )
    .infixLeft("=", Precedence.Assign, (left, _, right) => Assign(left, right, env))

    .construct()

export const execute: (command: string, env: UserEnvironmentView) => ExpressionResult = (
  str,
  env
) => calculator(env)(tokenize(str)).execute()
