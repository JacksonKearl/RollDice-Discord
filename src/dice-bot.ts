import { ParserBuilder, PARENTHESES_PARSELET, Tokenizer, TokenMatchResult } from "@jkearl/pratt"

import {
  Name,
  Number,
  Roll,
  Add,
  Sub,
  Mul,
  Advantage,
  DiceExpression,
  Neg,
  Div,
  Disadvantage,
  Bang,
  Assign,
  ExpressionResult
} from "./dice-expression"
import { UserEnvironmentView } from "./environment"

export const { tokenize } = new Tokenizer([
  { pattern: "+=" },
  { pattern: "-=" },
  { pattern: "+" },
  { pattern: "-" },
  { pattern: "*" },
  { pattern: "/" },
  { pattern: "(" },
  { pattern: ")" },
  { pattern: "!" },
  { pattern: "=" },
  { pattern: /@\s*a(d(v(a(n(t(a(g(e)?)?)?)?)?)?)?)?\b/, id: "ADVANTAGE" },
  { pattern: /@\s*d(i(s(a(d(v(a(n(t(a(g(e)?)?)?)?)?)?)?)?)?)?)?\b/, id: "DISADVANTAGE" },
  { pattern: /(\d+)?(?:d|D)(\d+)(k\d+)?/, id: "ROLL" },
  { pattern: /[a-zA-Z](\w|\.)*/, id: "NAME" },
  { pattern: /\d+/, id: "NUMBER" }
])

const enum Precedence {
  Assign = 1,
  At = 2,
  AddSub = 3,
  MulDiv = 4,
  Bang = 5,
  Negate = 6
}

export const calculator = (env: UserEnvironmentView) =>
  new ParserBuilder<DiceExpression, TokenMatchResult>(tokenize)
    .registerPrefix("NUMBER", { parse: (_, token) => new Number(token.value) })
    .registerPrefix("ROLL", { parse: (_, token) => new Roll(token.value) })
    .registerPrefix("NAME", { parse: (_, token) => new Name(token.value, execute, env) })
    .registerPrefix("(", PARENTHESES_PARSELET)

    .prefix("-", Precedence.Negate, (_, right) => Neg(right))

    .postfix("!", Precedence.Bang, left => Bang(left))

    .postfix("ADVANTAGE", Precedence.At, left => Advantage(left))
    .postfix("DISADVANTAGE", Precedence.At, left => Disadvantage(left))

    .infixLeft("/", Precedence.MulDiv, (left, _, right) => Div(left, right))
    .infixLeft("*", Precedence.MulDiv, (left, _, right) => Mul(left, right))
    .infixLeft("+", Precedence.AddSub, (left, _, right) => Add(left, right))
    .infixLeft("-", Precedence.AddSub, (left, _, right) => Sub(left, right))

    .infixLeft("+=", Precedence.Assign, (left, _, right) =>
      Assign(left, Add(Bang(left), right), env)
    )
    .infixLeft("-=", Precedence.Assign, (left, _, right) =>
      Assign(left, Sub(Bang(left), right), env)
    )
    .infixLeft("=", Precedence.Assign, (left, _, right) => Assign(left, right, env))

    .construct()

export const execute: (command: string, env: UserEnvironmentView) => ExpressionResult = (
  str,
  env
) => calculator(env)(str).execute()
