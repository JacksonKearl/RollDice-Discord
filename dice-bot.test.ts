import * as DiceBot from "./dice-bot"

describe("DiceBot", () => {
  describe("Tokenization", () => {
    it("can tokenize dice rolls", () => {
      let tokens = DiceBot.tokenize("d20")
      let executor = DiceBot.calculator({})(tokens)
      expect(executor).toMatchSnapshot()
    })
  })
})
