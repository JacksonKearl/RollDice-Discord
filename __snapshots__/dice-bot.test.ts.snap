// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiceBot Tokenization can tokenize dice rolls 1`] = `
Roll {
  "numDice": 1,
  "numKeep": 1,
  "numSides": 20,
  "rollString": "d20",
}
`;
